lib/linked_list.rb

```diff
@@ -12,150 +12,266 @@ def initialize(value, next_node = nil)
-    def find_max
-      raise NotImplementedError
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def find_max
+    return nil if !@head
+    current = @head
+    max = -(Float::INFINITY)
+    while current != nil
+      if current.data > max
+        max = current.data
+      end
+      current = current.next
     end
+    return max
+  end
 
-    # method to return the min value in the linked list
-    # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_min
-      raise NotImplementedError
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def find_min
+    return nil if !@head
+    current = @head
+    min = Float::INFINITY
+    while current != nil
+      if current.data < min
+        min = current.data
+      end
+      current = current.next
     end
+    return min
+  end
 
-
-    # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def length
-      raise NotImplementedError
+  # method that returns the length of the singly linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def length
+    counter = 0
+    current = @head
+    while current != nil
+      counter += 1
+      current = current.next
     end
+    return counter
+  end
 
-    # method that returns the value at a given index in the linked list
-    # index count starts at 0
-    # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def get_at_index(index)
-      raise NotImplementedError
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def get_at_index(index)
+    i = 0
+    current = @head
+    while current != nil
+      return current.data if index == i
+      i += 1
+      current = current.next
     end
+    return nil
+  end
 
-    # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def visit
-      raise NotImplementedError
+  # method to print all the values in the linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def visit
+    current = @head
+    while current != nil
+      puts current.data
+      current = current.next
     end
+  end
 
-    # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def delete(value)
-      raise NotImplementedError
+  # method to delete the first node found with specified value
+  # Time Complexity: O(1)
+  # Space Complexity: O(1)
+  def delete(value)

```

Comment:
👍 , However since this method traverses the list until it finds the given value and then deletes it.  So the time complexity is O(n)